File: frontend/src/utils/fileCategories.js
/**
 * What a file on a patient's record is, and therefore which tab shows it.
 *
 * Mirrors backend/core/file_categories.py. One list, so Imaging and Documents
 * can never both claim a file or both ignore it — which is exactly what used to
 * happen: every upload landed in Documents and Imaging was permanently empty.
 *
 * The words are the clinic's own and are stored verbatim, never translated on
 * display, so an export and the screen say the same thing.
 */

export const IMAGING_CATEGORIES = [
  'IOPA', 'RVG', 'OPG', 'Bitewing', 'CBCT', 'Photo', 'Scan',
];

export const DOCUMENT_CATEGORIES = [
  'Report', 'Referral', 'Estimate', 'Insurance', 'Consent', 'Other',
];

// What older builds send. Kept because an installed tablet keeps sending its
// own vocabulary for months, and a file filed under a word no screen filters on
// is a file nobody finds.
const ALIASES = {
  'x-ray': 'IOPA',
  xray: 'IOPA',
  'intra-oral': 'Photo',
  intraoral: 'Photo',
  periapical: 'IOPA',
  panoramic: 'OPG',
  // The old hardcoded server value. It means "nobody said", not a kind.
  document: '',
  upload: '',
};

const BY_LOWER = Object.fromEntries(
  [...IMAGING_CATEGORIES, ...DOCUMENT_CATEGORIES].map((c) => [c.toLowerCase(), c])
);

/** One canonical spelling, or '' when nothing usable was stored. */
export const normaliseCategory = (value) => {
  if (typeof value !== 'string') return '';
  const key = value.trim().toLowerCase();
  if (!key) return '';
  if (key in ALIASES) return ALIASES[key];
  return BY_LOWER[key] || '';
};

/** True when this file belongs on the Imaging tab. */
export const isImaging = (value) => IMAGING_CATEGORIES.includes(normaliseCategory(value));

// Tint per imaging type, so a badge and its filter chip cannot drift apart.
export const IMAGING_STYLE = {
  IOPA: 'bg-[#2a276e]/[0.08] text-[#2a276e]',
  RVG: 'bg-sky-50 text-sky-700',
  OPG: 'bg-violet-50 text-violet-700',
  Bitewing: 'bg-blue-50 text-blue-700',
  CBCT: 'bg-amber-50 text-amber-700',
  Photo: 'bg-emerald-50 text-emerald-700',
  Scan: 'bg-gray-100 text-gray-600',
};

File: backend/core/file_categories.py
"""What a file uploaded to a patient's record actually is.

Two screens show a patient's files: Imaging (films, photos, scans) and
Documents (referrals, reports, estimates, consents). Which one a file belongs on
is decided by this single list, so the two can never both claim it or both
ignore it.

─── Why this exists ─────────────────────────────────────────────────────────

Every upload landed in Documents and Imaging was permanently empty. Three
separate causes, and each on its own was enough:

  1. The upload endpoint did not accept a category at all, so the picker on the
     upload drawer wrote nothing anywhere.
  2. The list endpoint hardcoded `category="document"` on every row, so even a
     stored category was discarded on the way out.
  3. The Imaging tab read `xray_images`, a different table, which nothing the
     clinic could reach ever wrote to — `POST /xray/upload` rejects anything
     that is not a `.dcm` and saves to container-local disk, so it was never
     usable for a JPG off a sensor and would not have survived a deploy anyway.

So imaging is a CATEGORY OF DOCUMENT, stored in R2 like every other file in the
app, and the Imaging tab is a filtered view of the same table. The legacy
`xray_images` rows are still read and shown; nothing new is written there.
"""

# Films, photographs and scans — everything that belongs on the Imaging tab.
# The vocabulary is the clinic's own and is stored verbatim, never translated on
# display, so an export and the screen say the same word.
IMAGING_CATEGORIES = (
    'IOPA',      # single periapical film
    'RVG',       # radiovisiography sensor capture
    'OPG',       # panoramic
    'Bitewing',
    'CBCT',
    'Photo',     # intra-oral or clinical photograph
    'Scan',      # an intra-oral scanner's output, or anything else imaged
)

# Paperwork — everything that belongs on the Documents tab.
DOCUMENT_CATEGORIES = (
    'Report',
    'Referral',
    'Estimate',
    'Insurance',
    'Consent',
    'Other',
)

ALL_CATEGORIES = IMAGING_CATEGORIES + DOCUMENT_CATEGORIES

# What older clients send, and what it means now. The upload drawer offered
# 'X-Ray' and 'Intra-Oral' and threw both away; installed tablets will keep
# sending them for a while, and a file filed under a word no screen filters on
# is a file nobody finds.
LEGACY_CATEGORY_ALIASES = {
    'x-ray': 'IOPA',
    'xray': 'IOPA',
    'intra-oral': 'Photo',
    'intraoral': 'Photo',
    'periapical': 'IOPA',
    'panoramic': 'OPG',
    'document': None,   # the old hardcoded value: means "nobody said", not a kind
    'upload': None,
}

_BY_LOWER = {c.lower(): c for c in ALL_CATEGORIES}


def normalise_category(value):
    """One canonical spelling, or None when nothing usable was sent.

    None is a real answer and is kept as one: a file uploaded before the picker
    existed has no category, and guessing 'Other' for it would be inventing a
    fact. Both screens treat an uncategorised upload as paperwork, which is
    where they have always appeared.
    """
    if not isinstance(value, str):
        return None
    key = value.strip().lower()
    if not key:
        return None
    if key in LEGACY_CATEGORY_ALIASES:
        return LEGACY_CATEGORY_ALIASES[key]
    return _BY_LOWER.get(key)


def is_imaging(value) -> bool:
    """True when this category belongs on the Imaging tab."""
    return normalise_category(value) in IMAGING_CATEGORIES
